rfc/rfc1-game-engine-initial-design.md

```diff
@@ -0,0 +1,60 @@
+# Game engine initial design
+
+## Information needed by the client
+
+The client will need to have the following information at all times during a game. 
+
+For all game stages:
+- Current game stage (i.e. bidding or main game)
+- Current player
+- Remaining time to play/bid for the current player
+
+For the bidding stage:
+- Current highest bidding (i.e. the last one)
+    - Suit
+    - Value
+    - Whether there is "coinche" (i.e. counter) or "sur-coinche" (i.e. double-counter).
+- List of possible bidding for the current player
+      
+For the main game stage:
+- Cards already played for the current trick
+- Cards in the hand of the player
+- Number of cards in the hand of the other players
+- Cards the player is authorised to play
+- Trump suit (for display purposes only)
+
+## Actions we need to be able to perform from the client
+
+For the bidding stage:
+- make a bidding (including coinche and sur-coinche)
+- pass
+
+For the main game stage:
+- play an authorised card
+
+## Design for the game engine
+
+We need an engine able to cope with the constraints/rules detailed above.
+The game engine has the following properties:
+- An encapsulated game state to store all the game information
+- Getters to retrieve useful information about the game (listed in the previous section).
+- A way to make the state evolve as the game is played.
+
+**State representation:**
+The current game state is stored so that the information listed above is easily accessible (see following diagram).
+
+**Modification of the state:**
+Each state modification corresponds to an event:
+- An event is a simple object with a name, a timestamp, a playerId (when relevant) and some additional attributes. It represents something that occurred. 
```

Comment:
- An event corresponds to something (for now a player action only, but could very well be a disconnection event, a player sending a message in a chat, a commit in git...) that happened once and for all, and can't be changed. For this reason, it makes sense to associate an event with a timestamp (it's usually the case in event-driven systems).
- In our example, we can use this timestamp to know when a player played a card to compute how much time is remaining for the following player to play his card.

Reply:
I'm not sure we should mix apples and oranges and have events for both game moves and chat messages or other things not related to the actual game. It's likely better if we decouple the game engine from everything else, so that we can easily change the different parts of the system in the future.

It could indeed be useful for the timer. Let's have timestamps for now, as it'll be easier to remove them than add them in the future.